// lab/lib/reporters/console.js
// Load modules

var Tty = require('tty');
var Diff = require('diff');


// Declare internals

var internals = {
    width: 50
};


exports = module.exports = internals.Reporter = function (options) {

    this.settings = options;
    this.count = 0;
    this.last = [];

    this.colors = internals.colors(options.colors);
};


internals.Reporter.prototype.start = function (notebook) {

};


internals.Reporter.prototype.test = function (test) {

    if (this.settings.progress === 0) {
        return;
    }

    if (this.settings.progress === 1) {

        // ..x....x.-..

        if (!this.count) {
            this.report('\n  ');
        }

        this.count++;

        if ((this.count - 1) % internals.width === 0) {
            this.report('\n  ');
        }

        this.report(test.err ? this.colors.red('x') : (test.skipped || test.todo ? this.colors.magenta('-') : '.'));
    }
    else {

        var check = process.platform === 'win32' ? '\u221A' : '\u2714';
        var asterisk = process.platform === 'win32' ? '\u00D7' : '\u2716';

        // Verbose (Spec reporter)

        for (var i = 0, il = test.path.length; i < il; ++i) {
            if (test.path[i] !== this.last[i]) {
                this.report(internals.spacer(i * 2) + test.path[i] + '\n');
            }
        }

        this.last = test.path;

        var spacer = internals.spacer(test.path.length * 2);
        if (test.err) {
            this.report(spacer + this.colors.red(asterisk + test.id + ') ' + test.relativeTitle) + '\n');
        }
        else {
            var symbol = test.skipped || test.todo ? this.colors.magenta('-') : this.colors.green(check);
            this.report(spacer + symbol + ' ' + this.colors.gray(test.id + ') ' + test.relativeTitle +
                ' (' + test.duration + ' ms)') + '\n');
        }
    }
};


internals.spacer = function (length) {

    return new Array(length + 1).join(' ');
};


internals.Reporter.prototype.end = function (notebook) {

    if (this.settings.progress) {
        this.report('\n\n');
    }

    // Colors

    var red = this.colors.red;
    var redBg = this.colors.redBg;
    var green = this.colors.green;
    var greenBg = this.colors.greenBg;
    var magenta = this.colors.magenta;
    var gray = this.colors.gray;
    var yellow = this.colors.yellow;

    // Tests

    var failures = notebook.tests.filter(function (test) {

        return !!test.err;
    });

    var skipped = notebook.tests.filter(function (test) {

        return test.skipped;
    });

    var output = '';
    var totalTests = notebook.tests.length - skipped.length;

    var errors = notebook.errors || [];
    if (errors.length) {
        output += 'Test script errors:\n\n';
        for (var j = 0, jl = errors.length; j < jl; ++j) {
            var error = errors[j];
            output += red(error.message || error) + '\n';

            if (error.stack) {
                var stack = error.stack.slice(error.stack.indexOf('\n') + 1)
                                       .replace(/^/gm, '  ')
                                       .split('\n')
                                       .filter(internals.filterNodeModules)               // Remove node_modules files
                                       .slice(0, 5)                                       // Show only first 5 stack lines
                                       .join('\n');

                output += gray(stack) + '\n';
            }

            output += '\n';
        }
        output += red('There were ' + errors.length + ' test script error(s).') + '\n\n';
    }

    if (failures.length) {
        output += 'Failed tests:\n\n';

        for (var i = 0, il = failures.length; i < il; ++i) {
            var test = failures[i];
            var message = test.err.message || '';

            output += '  ' + test.id + ') ' + test.title + ':\n\n';

            // Actual vs Expected

            if (test.err.actual !== undefined &&
                test.err.expected !== undefined) {

                var actual = JSON.stringify(test.err.actual, null, 2);
                var expected = JSON.stringify(test.err.expected, null, 2);

                output += '      ' + redBg('actual') + ' ' + greenBg('expected') + '\n\n      ';

                var comparison = Diff.diffWords(actual, expected);
                for (var c = 0, cl = comparison.length; c < cl; ++c) {
                    var item = comparison[c];
                    var value = item.value;
                    var lines = value.split('\n');
                    for (var l = 0, ll = lines.length; l < ll; ++l) {
                        if (l) {
                            output += '\n      ';
                        }

                        if (item.added || item.removed) {
                            output += item.added ? greenBg(lines[l]) : redBg(lines[l]);
                        }
                        else {
                            output += lines[l];
                        }
                    }
                }

                output += '\n\n      ' + yellow(message);
                output += '\n\n';
            }
            else {
                output += '      ' + red(message) + '\n\n';
            }

            if (test.err.at) {
                output += gray('      at ' + test.err.at.filename + ':' + test.err.at.line + ':' + test.err.at.column) + '\n';
            }
            else if (!test.timeout &&
                test.err.stack) {

                output += gray(test.err.stack.slice(test.err.stack.indexOf('\n') + 1).replace(/^/gm, '  ')) + '\n';
            }

            output += '\n';
        }

        output += '\n' + red(failures.length + ' of ' + totalTests + ' tests failed') + '\n';
    }
    else {
        output += green(totalTests + ' tests complete') + '\n';
    }

    output += 'Test duration: ' + notebook.ms + ' ms\n';

    // Assertions

    if (notebook.assertions !== undefined) {
        output += 'Assertions count: ' + notebook.assertions + ' (verbosity: ' + (notebook.assertions / totalTests).toFixed(2) + ')\n';
    }

    // Leaks

    if (notebook.leaks) {
        if (note
